correct_periscope_drift.py

```diff
@@ -0,0 +1,431 @@
+    if len(evt_ra) != len(evt_dec):
+        raise ValueError("len(evt_ra) != len(evt_dec), {} != {}".format(
+                len(evt_ra), len(evt_dec)))
+
+    # Transform to Earth Centered Inertial
+    eci = radec2eci(evt_ra, evt_dec)
+    # transform from 3 x N to N x 3
+    eci = eci.transpose()
+
+    att_stack = np.repeat(np.array([ra_nom, dec_nom, roll_nom]),
+                          len(evt_ra)).reshape(3, len(evt_ra))
+    # Transforms
+    Ts = equatorial2transform(att_stack[0], att_stack[1], att_stack[2])
+
+    # The position of the events rotated into the frame of the
+    d_aca = np.sum(Ts.transpose(2, 0, 1) * eci, axis=-1).transpose()
+
+    R2A = 206264.81
+    yag = np.arctan2(d_aca[:, 1], d_aca[:, 0]) * R2A
+    zag = np.arctan2(d_aca[:, 2], d_aca[:, 0]) * R2A
+    return yag, zag
+
+
+def time_bins(times, x, nbins=20):
+    h, bins = np.histogram(times, bins=nbins)
+    bin_centers = (bins[:-1] + bins[1:]) / 2.0
+    inds = np.digitize(times, bins) - 1
+    bin_x = []
+    bin_std = []
+    for idx in range(0, nbins):
+        data = x[inds == idx]
+
+        bin_x.append(np.mean(data))
+        bin_std.append(np.std(data)/np.sqrt(len(data)))
+    return bin_centers, np.array(bin_x), np.array(bin_std)
+
+
+def fit(fit_data, evt_times, data_id, opt):
+
+    init_error = 5
+
+    ui.clean()
+    ui.load_arrays(data_id, evt_times - evt_times[0], fit_data,
+                   np.zeros_like(fit_data) + init_error)
+    v2("Fitting a line to the data to get reduced stat errors")
+    # First just fit a line to get reduced errors on this set
+    ui.polynom1d.line
+    ui.set_model(data_id, 'line')
+    ui.thaw('line.c1')
+    ui.fit(data_id)
+    fit = ui.get_fit_results()
+    calc_error = init_error * np.sqrt(fit.rstat)
+    ui.set_staterror(data_id, calc_error)
+    # Then fit the specified model
+    v2("Fitting a polynomial of degree {} to the data".format(opt['corr_poly_degree']))
+    ui.polynom1d.fitpoly
+    ui.freeze('fitpoly')
+    # Thaw the coefficients requested by the degree of the desired polynomial
+    for deg in range(1, 1 + opt['corr_poly_degree']):
+        ui.thaw("fitpoly.c{}".format(deg))
+    if opt['corr_poly_degree'] > 1:
+        ui.thaw('fitpoly.offset')
+    ui.set_model(data_id, 'fitpoly')
+    ui.fit(data_id)
+    mp = ui.get_model_plot(data_id)
+    return mp
+
+
+# The '@handle_ciao_errors' decorator will catch any error thrown and
+# display it in a format like that of a CIAO tool, then exit with a
+# non-zero status. So, if at any point in your code you need to exit just
+# raise an error and let handle_ciao_errors bother with the display.
+#
+# This also means that routines can be used from ChIPS/Sherpa/other scripts
+# without having to worry about them calling sys.exit.
+
+
+@handle_ciao_errors(TOOLNAME, VERSION)
+def main(opt):
+    events = extract_events(opt['evtfile'],
+                            opt['x'], opt['y'], opt['radius'])
+
+    evt_ra_nom = events.get_key('RA_NOM').value
```

Comment:
Just to clarify

There's reprojecting the files to a new tanget point (ie 'reproject_events') and 
then there is merging those files together (ie 'dmmerge') into a single evt file.

I could envision someone trying to run the tool on reprojected files which will 
have (generally) _NOM != _PNT.

Definitely agree ... running on a merged dataset is not something I would 
advocate/advise.  [FWIW:  The _NOM values should be the common tangent plane 
point.  Though if user merged without reprojecting, then they could be simply 
omitted/skipped if they are different, which is what happens to the _PNT values, 
since they are/will-be different, they will be omitted.]  And of course, all 
bets are off with the ROLL values (not affected by reproject'ing but are almost 
guaranteed to be different in merged and thus both ROLL_NOM & _PNT will be omitted).








On 06/21/16 16:00, Jean Connelly wrote:
>
> In correct_periscope_drift.py 
> <https://github.com/sot/correct_periscope_drift/pull/1#discussion_r67940644>:
>
> > +
> > +# The '@handle_ciao_errors' decorator will catch any error thrown and
> > +# display it in a format like that of a CIAO tool, then exit with a
> > +# non-zero status. So, if at any point in your code you need to exit just
> > +# raise an error and let handle_ciao_errors bother with the display.
> > +#
> > +# This also means that routines can be used from ChIPS/Sherpa/other scripts
> > +# without having to worry about them calling sys.exit.
> > +
> > +
> > +@handle_ciao_errors(TOOLNAME, VERSION)
> > +def main(opt):
> > +    events = extract_events(opt['evtfile'],
> > +                            opt['x'], opt['y'], opt['radius'])
> > +
> > +    evt_ra_nom = events.get_key('RA_NOM').value
>
> For this work, I think either Ra/Dec point (PNT or NOM) then is fine. The tool 
> is calculating corrections about a mean in the converted frame and then 
> applying the those corrections as additive offsets to the dy/dz values, so our 
> choice of the reference point here shouldn't be crucial either (it just biases 
> the mean)... BUT we need/want the roll to line up with something that was 
> close to the spacecraft roll (with the offset between the aspect camera frame 
> and the MNC frame being tiny enough to not be bothered with in this first draft).
>
> We need the aspect solution anyway, so I could calculate my own reference 
> point from the solution itself, but again, the differences shouldn't really 
> matter for this work.
>
> Also, I'm not really entertaining the idea of using this tool on merged 
> datasets, but what do those the ROLL_NOM values end up as?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sot/correct_periscope_drift/pull/1/files/cbe3d98891726c3e493d2afe4b325095778bf928#r67940644>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AGWoI6dt1D_IfPvFYGj3RTsj_wj6lrWJks5qOELdgaJpZM4IyK7D>.
>